src/NewKeyManager.sol

```diff
@@ -0,0 +1,186 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.8;
+
+import {IAndromeda} from "src/IAndromeda.sol";
+import {Secp256k1} from "src/crypto/secp256k1.sol";
+import {PKE, Curve} from "src/crypto/encryption.sol";
+import {BIP32} from "src/BIP32.sol";
+
+abstract contract NewKeyManagerBase {
+    // This base class provides the functionality of a singleton
+    // Private Key holder. It allows many applications to share the
+    // same bootstrapped instance.
+
+    // Anyone can see the master public key
+    address public xPub;
+
+    // placeholder for hardened and non_hardened child indexes
+    uint32 public constant HARDENED_START_INDEX = 2147483648;
+
+    // Attestation is now possible using the caller as domain
+    // separator
+    function attest(bytes32 appData) public returns (bytes memory) {
+        bytes32 digest = keccak256(abi.encodePacked(msg.sender, appData));
+        
+        return Secp256k1.sign(uint256(xPriv()), digest);
+    }
+
+    function verify(address caller, bytes32 appData, bytes memory sig) public view returns (bool) {
+        bytes32 digest = keccak256(abi.encodePacked(caller, appData));
+        return Secp256k1.verify(xPub, digest, sig);
+    }
+
+    //////////////////////////////////
+    // To be implemented by subclasses
+    //////////////////////////////////
+
+    // Only the contract in confidential mode can access the
+    // master private key
+    function seed() internal virtual returns (bytes32);
+
+    function xPriv() internal virtual returns (bytes32);
+
+    function _derivedPriv(address a, uint32 index) internal virtual returns (bytes32);
+
+    // Key derivation for encryption
+
+    // Because we are using hardened derivation, for each
+    // contract we will need someone to sign it off chain
+
+    mapping(bytes32 => bytes) internal _derivedPub;
+
+    function onchain_DeriveKey(
+        address a,
+        bytes memory dPub,
+        // Signature from a valid kettle
+        bytes memory sig,
+        uint32 index
+    ) public {
+        bytes32 digest = keccak256(abi.encodePacked(a, index, dPub));
+        require(Secp256k1.verify(xPub, digest, sig));
+        bytes32 key = keccak256(abi.encodePacked(a, index));
+        _derivedPub[key] = dPub;
+    }
+
+    function offchain_DeriveKey(address a, uint32 index) public returns (bytes memory dPub, bytes memory sig) {
+        bytes32 dPriv = _derivedPriv(a, index);
+        dPub = PKE.derivePubKey(dPriv);
```

Comment:
We included the indexing as a layer 2 in the key derivation tree.
So currently,
case1: if you don't pass an index (i.e. index=0), then the dPriv will be derived for the address directly from the BIP32 main extended private key.
The index of the address itself is being calculated like this:  hash(address) | 2³² to ensure that it derives a hardened key from the BIP32 library .

Case2: If we pass an index, then we allow to create a hardened child keypair for that specific address as you can see below in line [96](https://github.com/flashbots/andromeda-sirrah-contracts/pull/37/files#diff-ec66d355c9386f8d447752a90eedd71dbf337fb1809965136fe3429c0a9be221R96).

This way, for some use cases that @Ruteri mentioned to me where we want to create two private keys for an address, this would allow that.